### PR TITLE
Fix bug in `quantum_info.linalg_operations.matrix_power` for negative powers of singular matrices

### PR DESCRIPTION
**Pull request by renatomello**
Checklist:
- [x] Reviewers confirm new code works as expected.
- [x] Tests are passing.
- [x] Coverage does not decrease.
- [x] Documentation is updated.

### CODE REVIEW

**Comment by renatomello** [comment]
This is also ready and it is blocking #1470

File: src/qibo/backends/numpy.py [hunk 1]
```diff
@@ -769,12 +769,31 @@ def calculate_matrix_exp(self, a, matrix, eigenvectors=None, eigenvalues=None):
...
         if not isinstance(power, (float, int)):
             raise_error(
                 TypeError,
                 f"``power`` must be either float or int, but it is type {type(power)}.",
             )
+
+        if power < 0.0:
+            # negative powers of singular matrices via SVD
+            # det with np instead of self.np because of torch
+            determinant = np.linalg.det(matrix)
```
Comment (BrunoLiegiBastonLiegi) [inline src/qibo/backends/numpy.py:787]: is this going to work with `CupyBackend`? 
But in general why don't we use the usual `self.np`?

File: src/qibo/backends/pytorch.py [hunk 2]
```diff
@@ -199,9 +201,29 @@ def calculate_matrix_exp(self, a, matrix, eigenvectors=None, eigenvalues=None):
...
+        power: Union[float, int],
+        precision_singularity: float = 1e-14,
+    ):
+        if power < 0.0:
+            # negative powers of singular matrices via SVD
+            determinant = self.np.linalg.det(matrix)
+            if abs(determinant) < precision_singularity:
+                U, S, Vh = self.np.linalg.svd(matrix)
+                # cast needed because of different dtypes
+                S = self.cast(S)
```
Comment (BrunoLiegiBastonLiegi) [inline src/qibo/backends/pytorch.py:216]: if you cast in the `NumpyBackend` as well all this part won't be needed right?

File: src/qibo/backends/numpy.py (hunk 1 above)
Comment (renatomello) [reply src/qibo/backends/numpy.py:787]: It is working locally

**Review by Simone-Bordoni — APPROVED** [review]

File: src/qibo/backends/numpy.py (hunk 1 above)
Comment (BrunoLiegiBastonLiegi) [reply src/qibo/backends/numpy.py]: > In this specific point of the code the matrix might be a torch.Tensor or np.ndarray

but why is that the case? I mean the `backend.calculate_matrix_power` method should expect native arrays of the backend as inputs, right?

File: src/qibo/backends/pytorch.py [hunk 3]
```diff
@@ -199,9 +201,15 @@ def calculate_matrix_exp(self, a, matrix, eigenvectors=None, eigenvalues=None):
...
-        copied = super().calculate_matrix_power(copied, power)
+    def calculate_matrix_power(
+        self,
+        matrix,
+        power: Union[float, int],
+        precision_singularity: float = 1e-14,
+    ):
+        copied = self.cast(matrix, copy=True)
+        copied = self.to_numpy(copied) if power >= 0.0 else copied.detach()
+        copied = super().calculate_matrix_power(copied, power, precision_singularity)
```
Comment (BrunoLiegiBastonLiegi) [inline src/qibo/backends/pytorch.py:212]: This is due to `torch` not having a `fractional_matrix_power` alternative, right? Or is it due to some of the other `linalg` operations? 
I it's the first case, then I'd say you need to copy and cast to `numpy` only for `power > 0.`

Comment (renatomello) [reply src/qibo/backends/pytorch.py:212]: TBH idk how `detach` works in `torch`, but detaching the original will not affect the gradients? If the answer is no, then yes I can change it.

Comment (BrunoLiegiBastonLiegi) [reply src/qibo/backends/pytorch.py:212]: No indeed, detach basically discards the gradients. But I am not sure why do yo need to detach them for `power > 0.`. All the linalg operations you perform in that case should be `torch` compatible and should not have problems with the gradients if I am not mistaken, or am I missing something?

Comment (renatomello) [reply src/qibo/backends/pytorch.py:212]: some (maybe all?) `numpy` and `scipy` functions do not work on `torch` tensors if `requires_grad=True`. This error shows up on tests all the time. In this case, it is because of `scipy.linalg.fractional_matrix_power`.

Comment (BrunoLiegiBastonLiegi) [reply src/qibo/backends/pytorch.py:212]: Sorry, I meant for `power < 0.` This part in practice
``` python
            # negative powers of singular matrices via SVD
            determinant = self.np.linalg.det(matrix)
            if abs(determinant) < precision_singularity:
                U, S, Vh = self.np.linalg.svd(matrix)
                # cast needed because of different dtypes in `torch`
                S = self.cast(S)
                S_inv = self.np.where(
                    self.np.abs(S) < precision_singularity, 0.0, S**power
                )
                return (
                    self.np.linalg.inv(Vh) @ self.np.diag(S_inv) @ self.np.linalg.inv(U)
                )
```

Comment (renatomello) [reply src/qibo/backends/pytorch.py:212]: The code only enters this section if `power < 0.0` *and* the matrix is singular. So, for non-singular matrices, the execution still goes through the `scipy` function and thus either still needs the `.detach` or needs to check the determinant at that stage. I chose to just detach every tensor because it seemed it'd be less expensive than calculating the determinant, but I didn't rigorously check it.

Comment (BrunoLiegiBastonLiegi) [reply src/qibo/backends/pytorch.py:212]: Ah I see you are right sorry, I missed the det check, which you cannot do outside of the `NumpyBackend` method, I mean you could do but it wouldn't make much sense.

**Review by BrunoLiegiBastonLiegi — APPROVED** [review]

File: src/qibo/backends/numpy.py [hunk 4]
```diff
@@ -769,12 +769,32 @@ def calculate_matrix_exp(self, a, matrix, eigenvectors=None, eigenvalues=None):
...
+            if abs(determinant) < precision_singularity:
+                U, S, Vh = self.np.linalg.svd(matrix)
+                # cast needed because of different dtypes in `torch`
+                S = self.cast(S)
+                S_inv = self.np.where(
+                    self.np.abs(S) < precision_singularity, 0.0, S**power
+                )
+                return (
+                    self.np.linalg.inv(Vh) @ self.np.diag(S_inv) @ self.np.linalg.inv(U)
+                )
```
Comment (stavros11) [inline src/qibo/backends/numpy.py:796]: I didn't check all the details, so maybe I am wrong, but it seems that maybe we could reduce code repetition by moving this to a function that accepts `self.np` as an argument and then reuse for tensorflow (passing `self.tf`) and maybe even for cupy in qibojit.

I mean something like
```py
def calculate_matrix_negative_power(mod, matrix, ...):
    determinant = mod.linalg.det(matrix)
    ...
```
It can be defined even outside the backend, but then it will need to be imported in the various places, otherwise as `@staticmethod` in the backend.

**Review by stavros11 — APPROVED** [review]
